app/src/main/java/org/mozilla/focus/settings/ManualAddSearchEngineSettingsFragment.java

```diff
@@ -65,18 +112,141 @@ public boolean onOptionsItemSelected(MenuItem item) {
+                Snackbar.make(that.getView(), R.string.search_add_confirmation, Snackbar.LENGTH_SHORT).show();
+                that.getFragmentManager().popBackStack();
+            } else {
+                showServerError(that);
+            }
+
+            that.setUiIsValidatingAsync(false, that.menuItemForActiveAsyncTask);
+            that.activeAsyncTask = null;
+            that.menuItemForActiveAsyncTask = null;
+        }
+
+        private void showServerError(final ManualAddSearchEngineSettingsFragment that) {
+            final ManualAddSearchEnginePreference pref = (ManualAddSearchEnginePreference) that.findPreference(
+                    that.getString(R.string.pref_key_manual_add_search_engine));
+            pref.setSearchQueryErrorText(that.getString(R.string.error_hostLookup_title));
+        }
+    }
+
+    @WorkerThread // makes network request.
+    private static boolean isValidSearchQueryURL(final String query) {
+        // TODO: we should share the code to substitute and normalize the search string (see SearchEngine.buildSearchUrl).
+        final String encodedTestQuery = Uri.encode("test");
+
+        // We enforce HTTPS because some search engines will redirect http queries to https, which our
+        // validation (a 200 check) will fail, and since most search engines should support HTTPS, we
+        // try to be safer by defaulting to it.
+        final String normalizedHttpsSearchURLStr = enforceHTTPS(UrlUtils.normalize(query));
+        final String searchURLStr = normalizedHttpsSearchURLStr.replaceAll("%s", encodedTestQuery);
+
+        final URL searchURL;
+        try {
+            searchURL = new URL(searchURLStr);
+        } catch (final MalformedURLException e) {
+            // Don't log exception to avoid leaking URL.
+            Log.d(LOGTAG, "Malformed URL: returning invalid search query");
+            return false;
+        }
+
+        HttpURLConnection connection = null;
+        try {
+            connection = (HttpURLConnection) searchURL.openConnection();
+            connection.setConnectTimeout(SEARCH_QUERY_VALIDATION_TIMEOUT_MILLIS);
+            connection.setReadTimeout(SEARCH_QUERY_VALIDATION_TIMEOUT_MILLIS);
+
+            // Checking for 200 is not perfect - what if a search engine redirects us? -
+            // but we're in a rush and it's good enough for now.
+            return connection.getResponseCode() == 200;
+
+        } catch (final IOException e) {
+            // Don't log exception to avoid leaking URL.
+            Log.d(LOGTAG, "Failure to get response code from server: returning invalid search query");
+            return false;
+        } finally {
+            if (connection != null) {
+                try {
+                    connection.getInputStream().close(); // HttpURLConnection.getResponseCode opens the InputStream.
+                } catch (final IOException e) { } // Whatever.
```

Comment:
\[PMD\] WARNING: Must handle exceptions

Reply:
logcat StrictMode yelled at me that the InputStream wasn't closed and so I assume disconnect doesn't do it.